pkg/config/migration/four-to-five.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2019 the Service Broker Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migration
+
+func serviceNames() []string {
+	return []string{
+		"BIGQUERY",
+		"BIGTABLE",
+		"CLOUDSQL_MYSQL",
+		"CLOUDSQL_POSTGRES",
+		"DATAFLOW",
+		"DATASTORE",
+		"DIALOGFLOW",
+		"FIRESTORE",
+		"ML_APIS",
+		"PUBSUB",
+		"SPANNER",
+		"STACKDRIVER_DEBUGGER",
+		"STACKDRIVER_MONITORING",
+		"STACKDRIVER_PROFILER",
+		"STACKDRIVER_TRACE",
+		"STORAGE",
+	}
+}
+
+func customPlanKeys() (out []string) {
+	for _, s := range serviceNames() {
+		out = append(out, s+"_CUSTOM_PLANS")
+	}
+
+	return
+}
+
+// MergeToServiceConfig merges the custom plans, enabled, bind defaults, and
+// provision defaults for every 4.x service into a single configuration option.
+func MergeToServiceConfig() Migration {
+	jst := JsTransform{
+		EnvironmentVariables: []string{"GSB_SERVICE_CONFIG"},
+		MigrationJs: `
```

Comment:
This makes me so nervous. Sorry if I've asked this before, but is this nested JS code tested somehow?

Reply:
The embedded JS should all be tested in the unit-tests. Each test runs the same thing through the go and the JS version of the test. In the case of JsTransform, those use the same logic but different manipulation functions.

In the JS case, the test case is loaded into the `properties` JS field like OpsMan would do, and the test JS is run. Then, the newly modified properties is turned into environment variables via `tile2env` which mimics the way opsman converts the configuration JSON into environment variables.

In the Go case, the test case properties are run through `tile2env` first, then the go version of the function gets run which can either be hand-crafted, or sets up a JS environment which works directly on the environment variable map.

This makes sure that a user running the migration on environment variables gets the same result as if they'd been using the tile directly and exercises both sets of accessor/mutation functions.

I started down the path of having two distinct routes through the code, one written in Go and one in JS, but the code complexity exploded trying to reconcile Tile property names, environment variables, and the different underlying structures. I took a step back and decided that what I really wanted was code that worked through an interface. My "interface" ended up being `lookupProp(envVar) -> (Object|null)`, `deleteProp(envVar)`, and `setProp(envVar, value)`. Those worked cleanly on Go maps, and also on the properties.

The nice part is after the 5.0 release, all of this can be deleted. In the meantime, I'm happy to add anything you think is missing because I know this is less than ideal...

Reply:
Understood, in that case I'm much less nervous about this. Thanks :)